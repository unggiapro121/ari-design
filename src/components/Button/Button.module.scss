@import '../../styles/variables';
@import '../../styles/mixins';

$border-size: 2px;

@mixin colorVariant(
  $background,
  $background-hover,
  $color: $background,
  $color-hover: $color
) {
  position: relative;
  background: $background;
  border-color: $background;
  color: $color;

  &:visited {
    color: $color;
  }

  &::after {
    background: $background;
  }

  &:hover:not([class*='__inactive']) {
    background-color: $background-hover;
    border-color: $background-hover;
    color: $color-hover;

    &::after {
      background: $background-hover;
    }
  }
}

@mixin hoverVariant($background, $color) {
  &:hover:not([class*='__inactive']) {
    @if $background != '' {
      background-color: $background;
    }

    @if $color != '' {
      color: $color;
    }
  }
}

.button {
  @include font-style(no1);

  line-height: 1;
  box-sizing: border-box;
  user-select: none;
  width: 300px;
  height: 48px;
  margin: 0;
  padding: 12px 50px;
  align-items: center;
  justify-content: center;
  text-transform: uppercase;
  display: flex;
  gap: 6px;
  flex-shrink: 1;
  cursor: pointer;
  border: none;
  background: transparent;
  white-space: nowrap;
  overflow: hidden;
  text-decoration: none;
  font-family: inherit;

  @include sm {
    width: 100%;
  }

  &Spinner {
    position: absolute;
    right: 0;
    font-size: 24px;
    margin-right: $spacing-xs;
  }

  &__red {
    @include colorVariant($color-red, $color-red-hover);
  }

  &__blue {
    @include colorVariant($color-blue, $color-blue-hover);
  }

  &__black {
    @include colorVariant($color-black, $color-black-hover);
  }

  &__cobalt {
    @include colorVariant($color-cobalt, $color-cobalt-hover);
  }

  &__white {
    @include colorVariant($color-white, $color-white);
  }

  &__inactive {
    opacity: 0.5;
    cursor: auto;
  }

  &__primary {
    @include hoverVariant('', $color-white);

    border: none;
    color: $color-white;

    &:visited {
      color: $color-white;
    }

    &.button__white {
      @include hoverVariant('', $color-black);

      color: $color-black;

      &:visited {
        color: $color-black;
      }
    }
  }

  &__secondary {
    @include hoverVariant($color-white, '');

    border-width: 2px;
    border-style: solid;
    background: $color-white;

    &.button__white {
      @include hoverVariant(transparent, '');

      color: $color-white;
      background: transparent;

      &:visited {
        color: $color-white;
      }
    }
  }

  &__tertiary {
    @include hoverVariant($color-white, $color-black);

    width: auto;
    padding: 12px 0;
    background: transparent;
    color: $color-black;

    &:where(a) {
      display: inline-flex;
    }

    &::after {
      content: '';
      position: absolute;
      bottom: 11px;
      width: 100%;
      height: 2px;
    }

    &.button__white {
      @include hoverVariant(transparent, '');

      color: $color-white;
    }

    &:hover:not([class*='__inactive']) {
      background-color: transparent;
    }
  }

  &__noFill {
    background-color: transparent;
  }
}
