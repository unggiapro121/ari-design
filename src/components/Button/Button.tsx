import { LoadingOutlined } from '@ant-design/icons';
import clsx from 'clsx';
import React from 'react';

import Icon from '../Icon/Icon';

import styles from './Button.module.scss';

export enum ButtonVariant {
  Primary = 'primary',
  Secondary = 'secondary',
  Tertiary = 'tertiary',
}

export enum ButtonColor {
  Red = 'red',
  Blue = 'blue',
  Cobalt = 'cobalt',
  Black = 'black',
  White = 'white',
}

export enum ButtonType {
  Button = 'button',
  Submit = 'submit',
  Reset = 'reset',
  Link = 'link',
}

export interface ButtonProps {
  /**
   * Text displayed inside of the button
   */
  label: string;
  /**
   * URL, only when `type` is  `"link"`
   */
  href?: string;
  /**
   * Icon displayed before the label inside of the button
   */
  iconPrefix?: typeof Icon;
  /**
   * Icon displayed after the label inside of the button
   */
  iconSuffix?: typeof Icon;
  /**
   * Display variant of the button eg. Primary / Secondary / Tertiary
   */
  variant: ButtonVariant | `${ButtonVariant}`;
  /**
   * Color of the button eg. Red / Blue / Cobalt / Black / White
   */
  color: ButtonColor | `${ButtonColor}`;
  /**
   * Include the background colour fill on the button, for use with `secondary` variant
   */
  bgFill?: boolean;
  /**
   * Type of the button eg. Button / Submit / Reset / Link
   * @default Button
   */
  type?: ButtonType | `${ButtonType}`;
  /**
   * Whether to display the button as inactive state (still clickable)
   */
  inactive?: boolean;
  /**
   * Additional class styles for the button
   */
  className?: string;
  /**
   * Additional styles for the button
   */
  style?: React.CSSProperties;
  /**
   * Button onClick event handler
   */
  onClick?: React.MouseEventHandler;
  /**
   * Id of form to associate this button with
   */
  formId?: string;
  /**
   * Whether to display loading spinner
   */
  loading?: boolean;
}

/**
 * Loading spinner for button loading state 
 */
const Spinner = () => (<LoadingOutlined className={styles.buttonSpinner} data-testid="button-spinner"/>);

/**
 * Ampol style button and variances
 * @see {@link ButtonProps} for available props
 */
const Button = ({
  label,
  href,
  variant,
  color,
  bgFill = true,
  type = ButtonType.Button,
  style,
  className,
  onClick,
  inactive,
  formId,
  iconPrefix,
  iconSuffix,
  loading,
}: ButtonProps) => {

  const buttonStyle = clsx([
    styles.button,
    styles[`button__${color}`],
    styles[`button__${variant}`],
    inactive && styles.button__inactive,
    !bgFill && styles.button__noFill,
    className,
  ]);

  if (type === ButtonType.Link) {
    return <a href={href} className={buttonStyle} style={style} data-testid="Button">{iconPrefix}{label}{iconSuffix}</a>;
  }

  return <>
    <button type={type} className={buttonStyle} style={style} onClick={onClick} form={formId} data-testid="Button">
      {iconPrefix}{label}{iconSuffix}{loading && <Spinner />}
    </button>
  </>;
};

export default Button;
