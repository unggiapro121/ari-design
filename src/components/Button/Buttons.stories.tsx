import { Meta, Story } from '@storybook/react';
import * as React from 'react';

import Icon from '../Icon/Icon';

import Button, { ButtonColor, ButtonProps, ButtonType, ButtonVariant } from './Button';

export default {
  title: 'UI/Buttons',
  component: Button,
  parameters: {
    zeplinLink: [
      {
        name: 'Primary',
        link: 'https://app.zeplin.io/styleguide/6180631f214952a779c2ea9d/components?coid=6182072f33764d816dadf423',
      },
      {
        name: 'Secondary',
        link: 'https://app.zeplin.io/styleguide/6180631f214952a779c2ea9d/components?coid=618207304d672681da268944',
      },
      {
        name: 'Tertiary',
        link: 'https://app.zeplin.io/styleguide/6180631f214952a779c2ea9d/components?coid=618207302daed67f7781bde9',
      },
    ],
  },
} as Meta<typeof Button>;

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = { label: 'Button Label', variant: ButtonVariant.Primary, color: ButtonColor.Red };
export const PrimaryWithIcon = Template.bind({});
PrimaryWithIcon.args = { label: 'Connect', variant: ButtonVariant.Primary, color: ButtonColor.Red, iconSuffix: <Icon url={'../../images/icons/icon-external-link.svg'} /> };
export const Secondary = Template.bind({});
Secondary.args = { label: 'Button Label', variant: ButtonVariant.Secondary, color: ButtonColor.Red };
export const Tertiary = Template.bind({});
Tertiary.args = { label: 'Button Label', variant: ButtonVariant.Tertiary, color: ButtonColor.Red };
export const ButtonAsLink = Template.bind({});
ButtonAsLink.args = {
  type: ButtonType.Link,
  label: 'Anchor Tag Button',
  href: 'https://ampol.com.au',
  variant: ButtonVariant.Primary,
  color: ButtonColor.Cobalt,
  onClick: undefined,
};
