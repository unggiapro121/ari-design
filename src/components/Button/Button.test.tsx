import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

import Button, { ButtonColor, ButtonType, ButtonVariant } from './Button';

describe('Button component', () => {
  test('does render correctly', () => {
    const button = renderer
      .create(
        <Button
          label="Test"
          variant={ButtonVariant.Primary}
          color={ButtonColor.Red}
        />,
      )
      .toJSON();

    expect(button).toMatchSnapshot();
  });

  describe('button element is of type', () => {
    test('button by default', () => {
      const { getByText } = render(
        <Button
          label="Test"
          variant={ButtonVariant.Primary}
          color={ButtonColor.Red}
        />,
      );

      expect(getByText('Test').getAttribute('type')).toBe('button');
    });

    test('button', () => {
      const { getByText } = render(
        <Button
          label="Test"
          variant={ButtonVariant.Primary}
          color={ButtonColor.Red}
          type={ButtonType.Button}
        />,
      );

      expect(getByText('Test').getAttribute('type')).toBe('button');
    });

    test('submit', () => {
      const { getByText } = render(
        <Button
          label="Test"
          variant={ButtonVariant.Primary}
          color={ButtonColor.Red}
          type={ButtonType.Submit}
        />,
      );

      expect(getByText('Test').getAttribute('type')).toBe('submit');
    });

    test('reset', () => {
      const { getByText } = render(
        <Button
          label="Test"
          variant={ButtonVariant.Primary}
          color={ButtonColor.Red}
          type={ButtonType.Reset}
        />,
      );

      expect(getByText('Test').getAttribute('type')).toBe('reset');
    });

    test('link', () => {
      const { getByRole } = render(
        <Button
          label="Test"
          href="test.com"
          variant={ButtonVariant.Primary}
          color={ButtonColor.Red}
          type={ButtonType.Link}
        />,
      );
      const anchor = getByRole('link');

      expect(anchor).toBeTruthy();
      expect(anchor.textContent).toEqual('Test');
      expect(anchor.getAttribute('href')).toEqual('test.com');
    });
  });

  it('renders the button with the given label', () => {
    const screen = render(
      <Button
        label="Test Button"
        variant={ButtonVariant.Primary}
        color={ButtonColor.Red}
      />,
    );
    const button = screen.getByText('Test Button');

    expect(button).toBeInTheDocument();
  });

  it('renders the button with an icon prefix and suffix', () => {
    const { container } = render(
      <Button
        label="Test Button"
        variant={ButtonVariant.Primary}
        color={ButtonColor.Red}
        iconPrefix={<span data-testid="icon-prefix" />}
        iconSuffix={<span data-testid="icon-suffix" />}
      />,
    );

    expect(
      container.querySelector('[data-testid="icon-prefix"]'),
    ).toBeInTheDocument();
    expect(
      container.querySelector('[data-testid="icon-suffix"]'),
    ).toBeInTheDocument();
  });

  it('renders the button with a link', () => {
    const { container } = render(
      <Button
        label="Test Button"
        variant={ButtonVariant.Primary}
        color={ButtonColor.Red}
        type={ButtonType.Link}
        href="https://example.com"
      />,
    );

    const link = container.querySelector('a');

    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', 'https://example.com');
  });

  it('triggers onClick event when clicked', () => {
    const handleClick = vi.fn();

    const screen = render(
      <Button
        label="Test Button"
        variant={ButtonVariant.Primary}
        color={ButtonColor.Red}
        onClick={handleClick}
      />,
    );

    const button = screen.getByText('Test Button');

    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('renders the button with loading state', () => {
    const screen = render(
      <Button
        label="Test Button"
        variant={ButtonVariant.Primary}
        color={ButtonColor.Red}
        loading={true}
      />,
    );

    const spinner = screen.getByTestId('button-spinner');

    expect(spinner).toBeInTheDocument();
  });

  it('applies custom className and style', () => {
    const { container } = render(
      <Button
        label="Test Button"
        variant={ButtonVariant.Primary}
        color={ButtonColor.Red}
        className="custom-class"
        style={{ backgroundColor: 'yellow' }}
      />,
    );

    const button = container.querySelector('.custom-class');

    expect(button).toBeInTheDocument();
    expect(button).toHaveStyle({ backgroundColor: 'yellow' });
  });
});
