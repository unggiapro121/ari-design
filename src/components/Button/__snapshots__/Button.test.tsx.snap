// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component > does render correctly 1`] = `
<button
  className="_button_f8f296 _button__red_f8f296 _button__primary_f8f296"
  data-testid="Button"
  type="button"
>
  Test
</button>
`;
